src/logic/community/ranking.ts

```diff
@@ -0,0 +1,88 @@
+import { ICommunityRankingComponent, CommunityRankingMetrics } from './types'
+import { AppComponents } from '../../types/system'
+
+type MetricWeight = {
+  metric: Exclude<keyof CommunityRankingMetrics, 'communityId'>
+  weight: number
+}
+
+type MetricWeights = Array<MetricWeight>
+
+const DEFAULT_METRIC_WEIGHTS: MetricWeights = [
+  { metric: 'eventsCount', weight: 0.5 },
+  { metric: 'hasThumbnail', weight: 1 },
+  { metric: 'photosCount', weight: 0.2 },
+  { metric: 'hasDescription', weight: 1 },
+  { metric: 'placesCount', weight: 0.2 },
+  { metric: 'newMembersCount', weight: 0.4 },
+  { metric: 'postsCount', weight: 0.2 },
+  { metric: 'streamsCount', weight: 0.2 },
+  { metric: 'eventsTotalAttendees', weight: 0.01 },
+  { metric: 'streamsTotalParticipants', weight: 0.01 }
+]
+
+export function createCommunityRankingComponent({
+  logs,
+  communitiesDb,
+  communityThumbnail
+}: Pick<AppComponents, 'logs' | 'communitiesDb' | 'communityThumbnail'>): ICommunityRankingComponent {
+  const logger = logs.getLogger('ranking-component')
+
+  async function calculateRankingScore(metrics: CommunityRankingMetrics): Promise<number> {
+    const weights: MetricWeights = DEFAULT_METRIC_WEIGHTS
+    const score = weights.reduce((sum, { metric, weight }) => sum + metrics[metric] * weight, 0)
+
+    // TODO: consider normalizing the score to a scale of 0 to 1
+    return Math.max(0, score)
+  }
+
+  async function calculateRankingScoreForAllCommunities(): Promise<void> {
+    logger.info('Starting ranking score calculation for all communities')
+
+    try {
+      // Get all communities with their ranking metrics calculated in a single query
+      const communitiesMetrics = await communitiesDb.getAllCommunitiesWithRankingMetrics()
+
+      if (communitiesMetrics.length === 0) {
+        logger.info('No active communities found for ranking calculation')
+        return
+      }
+
+      // Check thumbnails for all communities in a single batch
+      const communityIds = communitiesMetrics.map(({ communityId }) => communityId)
+      const thumbnails = await communityThumbnail.getThumbnails(communityIds)
+
+      for (const communityMetrics of communitiesMetrics) {
+        try {
+          const metrics: CommunityRankingMetrics = {
+            ...communityMetrics,
+            hasThumbnail: thumbnails[communityMetrics.communityId] ? 1 : 0
+          }
+          const score = await calculateRankingScore(metrics)
+
+          logger.info('Calculated ranking score for community', {
+            communityId: communityMetrics.communityId,
+            score
+          })
+
+          // Update ranking score
+          await communitiesDb.updateCommunity(communityMetrics.communityId, { ranking_score: score })
```

Comment:
What if we either:
- Store the `ranking_score` in the `community_ranking_metrics` when inserting or updating the table, instead of inserting it in the community's table. This would require a LEFT JOIN with the metrics.
- Compute the `ranking_score` when getting the communities by doing a LEFT JOIN with the `community_ranking_metrics` table.

Having the metrics data on a single table would simplify all the process of updating and computing them, specially the part of being consistent with the data in both tables. I would compute them separately as it's done here only if there's a performance issue.
What do you think about this approach?